Mathlib/Algebra/BigOperators/Order.lean

```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
 #align is_absolute_value.map_prod IsAbsoluteValue.map_prod
 
 end AbsoluteValue
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+
+/-- Prove that a finset is nonempty from a local assumption or from the fact that the finset is
+`univ` in a nonempty fintype.
+
+Note that this does not do any complicated reasoning: the only ways it can prove `s` is nonempty is
+if there is a local `s.Nonempty` hypothesis or `s = (univ : Finset α)` and `Nonempty α` can be
+synthesized by TC inference.
+
+TODO: Replace by an aesop ruleset? -/
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM (Option Q(Finset.Nonempty $s)) := do
+  if let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) then
+    return some (.fvar fv)
+  else if let ~q(@univ _ $fi) := s then
+    if let .some _no ← trySynthInstanceQ q(Nonempty $α) then
+      return some q(Finset.univ_nonempty)
+    else return none
+  else return none
```

Comment:
Thanks to Jannis' help, I came up with:
```suggestion
-/
def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
    MetaM (Option Q(Finset.Nonempty $s)) := do
  -- Aesop expects to operate on goals, so we're going to make a new goal.
  let goal ← Lean.Meta.mkFreshExprMVar q(Finset.Nonempty $s)
  let mvar := goal.mvarId!
  -- We want this to be fast, so use only the basic and `Finset.Nonempty` specific rules.
  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `Finset.Nonempty]
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ← Aesop.search mvar (.some rules)
  -- Fail if there are open goals remaining, this serves as an extra check for the
  -- Aesop configuration option `terminal := true`.
  if remainingGoals.size > 0 then return none
  Lean.getExprMVarAssignment? mvar
```

With the following extra declarations:
```lean
-- `Finset.Nonempty` rules try to prove that a given finset is nonempty,
-- for use in the `Finset.sum` positivity extension.
declare_aesop_rule_sets [Finset.Nonempty] (default := true)
```
and in a separate file:
```lean
attribute [aesop unsafe apply (rule_sets [Finset.Nonempty])] Finset.univ_nonempty
```
I would argue that the `declare_aesop_rule_sets` line should go into a new file (maybe `Mathlib/Data/Finset/Attr.lean`?) and the `attribute` can directly go onto the declaration of `Finset.univ_nonempty`.

Reply:
(It is still an annoyance that environment extensions only become visible after importing!)

Reply:
I'm putting it in `Data.Finset.Basic` for now. I will want it available for the `Finset.card` extension.